src/dbnode/storage/database.go

```diff
@@ -317,14 +309,61 @@ func (d *db) UpdateOwnedNamespaces(newNamespaces namespace.Map) error {
 			"restart the process if you want changes to take effect")
 	}
 
-	// enqueue bootstraps if new namespaces
 	if len(adds) > 0 {
-		d.queueBootstrapWithLock()
-	}
+		if noBootstrapNeeded {
+			d.Lock()
+			defer d.Unlock()
+			return d.addNamespacesWithLock(adds)
+		}
+
+		if mediator := d.mediator; mediator == nil || !mediator.IsOpen() {
```

Comment:
Hm this seems to be done out of a lock, however d.mediator can be set to nil yeah?

Perhaps there should be an extra method added to db called `mediatorIsOpen() bool` that takes an RLock, checks if nil and if not nil checks that it is open, that way this logic becomes safe (with the RLock) and also reuseable instead of repeated several times.